hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java

```diff
@@ -0,0 +1,249 @@
+    }
+
+    /**
+     * Removes all expired {@link TransactionID}s from the cache.
+     */
+    private void removeExpiredTransactions(@NonNull final WritableQueueState<TransactionRecordEntry> queue) {
+        // Compute the earliest valid start timestamp that is still within the max transaction duration window.
+        final var now = asTimestamp(Instant.now());
+        final var earliestValidState = minus(now, props.maxTxnDuration());
+
+        // Loop in order and expunge every entry where the timestamp is before the current time. Also remove from the
+        // in memory data structures.
+        final var itr = queue.iterator();
+        while (itr.hasNext()) {
+            final var entry = itr.next();
+            final var rec = entry.transactionRecordOrThrow();
+            final var txId = rec.transactionIDOrThrow();
+            // If the timestamp is before the current time, then it has expired
+            if (isBefore(txId.transactionValidStartOrThrow(), earliestValidState)) {
+                // Remove from the histories
+                itr.remove();
+                // Remove from the payer to transaction index
+                final var payerAccountId = txId.accountIDOrThrow(); // NOTE: Not accurate if the payer was the node
+                final var transactionIDs =
+                        payerToTransactionIndex.computeIfAbsent(payerAccountId, ignored -> new HashSet<>());
+                transactionIDs.remove(txId);
+                if (transactionIDs.isEmpty()) {
+                    payerToTransactionIndex.remove(payerAccountId);
+                }
+            } else {
+                return;
+            }
+        }
+    }
+
+    // ---------------------------------------------------------------------------------------------------------------
+    // Implementation methods of RecordCache
+    // ---------------------------------------------------------------------------------------------------------------
+
+    @NonNull
+    @Override
+    public List<TransactionRecord> getRecords(@NonNull final TransactionID transactionID) {
+        final var history = histories.get(transactionID);
+        return history == null ? emptyList() : history.records();
+    }
+
+    @NonNull
+    @Override
+    public List<TransactionRecord> getRecords(@NonNull final AccountID accountID) {
+        final var transactionIDs = payerToTransactionIndex.get(accountID);
+        if (transactionIDs == null) {
+            return emptyList();
+        }
+
+        // NOTE: We should add to our queries an upper limit in terms of the number of items to return in response.
+        // Since we have a limit of the total number of items that can be in memory (whatever the max throttle can do
+        // in `maxTxnDuration` seconds), there is a practical upper limit. That being said, it would be even better to
+        // limit this to a small number of transactions -- say, 50 -- and then have the client page through the results.
```

Comment:
Imo do need to enforce, since a query thread asking for an active payer account could be making tens of thousands of `histories.get()` calls here, causing heavy contention with `thread-cons`.

Reply:
Does our API permit this today, or are you saying that yes, this is something we should do, but not part of this PR?

Reply:
The current API does [limit the # of records returned](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/queries/answering/AnswerFunctions.java#L51) to the value of property [`ledger.records.maxQueryableByAccount=180`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties#L133).

Reply:
But not sure what I was thinking, `ConcurrentHashMap.get()` does not acquire a lock, so no contention issues. 😵

Reply:
OK, this is now implemented.